feat(lib): add resource move functionality (#3152)
hashicorp/terraform-cdk: website/docs/cdktf/concepts/resources.mdx

</CodeTabs>

> ## Refactoring & Renaming Resources

Reviewer: We have a guide on refactoring, we should update it using the move block. (We should also write one for import, if we have time)

Author: Great catch, if anything that section logically belongs in our refactoring guide.

[Moved the bulk of the section](https://github.com/hashicorp/terraform-cdk/pull/3152/commits/9a29db9c7fc3c16d444e7fc727101b56ac829af0) to the refactoring guide. I then have `resources.mdx` mention the functionality and refer to the refactoring guide for more info.

</CodeTabs>

## Refactoring & Renaming Resources

When working with your infrastructure definitions and the need arises to refactor or rename resources without destroying and recreating them, you can leverage the `moveTo` function. This section outlines how to perform resource moves, rename resources, and enable foreach on existing resources in your CDKTF project.

### Performing Resource Moves

> The `moveTo` function is available on all resources and is used for relocating a resource to the location specified by the string target `moveTarget`. To set the `moveTarget` for a resource, utilize the `addMoveTarget` function that is present on the resource to move to.

Reviewer 2: Maybe add something along these lines?
```suggestion
The `moveTo` function is available on all resources and is used for relocating a resource to the location specified by the string target `moveTarget`. To set the `moveTarget` for a resource, utilize the `addMoveTarget` function that is present on the resource to move to. The `moveTarget` can be a string of your choice but has to be unique within your stack.
```

Author: included 👍

> The `moveTo` function is available on all resources and is used for relocating a resource to the location specified by the string target `moveTarget`. To set the `moveTarget` for a resource, utilize the `addMoveTarget` function that is present on the resource to move to.

Reviewer 3: ```suggestion
The `moveTo` function is available on all resources and is used for relocating a resource to the location specified by the string target `moveTarget`. To set a resource's `moveTarget`, use the `addMoveTarget` function that is present on the resource to move to. You can specify an arbitrary string for the `moveTarget`, but it must be unique within your stack.
```
I'm having trouble parsing "use the `addMoveTarget` function that is present on the resource to move to". It might be helpful to use a more concrete verb to describe what you do with the `addMoveTarget` function, for example "add the `addMoveTarget` function to the destination resource. Terraform moves the resource configured with the `moveTo` function to the target during the `apply` operation." I just made that up, of course, but it I think it would be helpful to include that level of (accurate) detail.

Author: Huge improvement on that wording, [included 👍].(https://github.com/hashicorp/terraform-cdk/pull/3152/commits/70e4146b79b73f51e3b568df7fd5918366cb5b49)

```ts
new S3Bucket(this, "test-bucket-move-to", {
  bucket: "move-bucket-name",
}).addMoveTarget("move-s3");

new S3Bucket(this, "test-bucket-move-from", {
  bucket: "move-bucket-name",
}).moveTo("move-s3");
```

After deployment, both the resource being moved and the move target on the destination resource can be removed.

#### Move Targets

> A `moveTarget` is accessible anywhere within the context of the stack where it is created. Whether it's set at the root of the stack or within a nested construct, you can access it accordingly.

Reviewer 3: ```suggestion
A `moveTarget` is accessible anywhere within the context of the stack where it is created, including the root of the stack and within a nested construct. Moving resources to a different stack is not supported. 
```
Is that more concise? Per the style guide, temporal language, such as "in a future version . . .", "at this time", etc. should be omitted.

Author: Will note to not use such language in the future, [included](https://github.com/hashicorp/terraform-cdk/pull/3152/commits/70e4146b79b73f51e3b568df7fd5918366cb5b49) 👍

> A `moveTarget` is accessible anywhere within the context of the stack where it is created. Whether it's set at the root of the stack or within a nested construct, you can access it accordingly.

Author: included 👍

```ts
new S3Bucket(this, "test-bucket-1", {
  bucket: "test-move-bucket-name-1",
}).moveTo("move");

// Contains resource with moveTarget "move"
new NestedConstructToMoveTo(this, "construct-to-move-to");
```

### Renaming a Resource

To rename a resource, you can simply use the `renameResourceId` function available on the resource.

```ts
new S3Bucket(this, "bucket-id", {
  bucket: "bucket-rename",
}).renameResourceId("bucket-new-id");
```

After deploying the resource, you can update your resource to use the newly assigned resource ID, and the `renameResourceId` call can be safely removed.

### Enabling `foreach` on a Resource

> If you wish to incorporate a deployed resource into a composition using `foreach`, you can achieve this without destroying the resource by specifying the desired index using the `moveTo` function. The index should correspond to the key found in the `TerraformIterator` named iterator

Reviewer 2: ```suggestion
If you wish to incorporate a deployed resource into a composition using `foreach`, you can achieve this without destroying the resource by specifying the desired index as a second argument for the `moveTo` function. The index should correspond to the key found in the `TerraformIterator` named `iterator`
```

Reviewer 2: Not to happy with my "for the moveTo function" wording, maybe you have some better idea to include that there?
(I'd like to make clear that the index is not an alternative to the moveTarget but rather an additional param)

Author: included 👍

> If you wish to incorporate a deployed resource into a composition using `foreach`, you can achieve this without destroying the resource by specifying the desired index using the `moveTo` function. The index should correspond to the key found in the `TerraformIterator` named iterator

Reviewer 3: ```suggestion
To incorporate a deployed resource into a composition without destroying the resource, use the `foreach` function and specify an index for the `moveTo` function. The index should correspond to the key in the `TerraformIterator` named `iterator`. 
```
Is that clearer?

Author: Far clearer, changed your suggestion to use some more correct wording but otherwise [included](https://github.com/hashicorp/terraform-cdk/pull/3152/commits/70e4146b79b73f51e3b568df7fd5918366cb5b49) 👍

</CodeTabs>

## Refactoring & Renaming Resources

> By default moving and/or renaming resource will be regarded with the intent to destroy the resource and re-create it. If such behavior must be avoided, but there is need to refactor your infrastructure definitions and/or rename resources in it, the function `moveTo` can serve your purpose.

Reviewer 2: ```suggestion
By default moving and/or renaming resources will be regarded with the intent to destroy the resource and re-create it. If such behavior must be avoided, but there is need to refactor your infrastructure definitions and/or rename resources in it, the function `moveTo` can serve your purpose.
```

Also: It seems like there's quite a bit of passive voice that I learned we usually try to avoid when writing docs.
